libkbfs/block_retrieval_queue.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"container/heap"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultBlockRetrievalWorkerQueueSize int = 100
+	defaultOnDemandRequestPriority       int = 100
+)
+
+// blockRetrievalRequest represents one consumer's request for a block.
+type blockRetrievalRequest struct {
+	ctx    context.Context
+	block  Block
+	doneCh chan error
+}
+
+// blockRetrieval contains the metadata for a given block retrieval. May
+// represent many requests, all of which will be handled at once.
+type blockRetrieval struct {
+	// The block being retrieved
+	blockPtr BlockPointer
+	// The key metadata for the request
+	kmd KeyMetadata
+	// The index of the retrieval in the heap
+	index int
+	// The priority of the retrieval
+	priority int
+	// The count of elements in the heap when this element was inserted.
+	// Maintains FIFO.
+	insertionOrder uint64
+	// All the individual requests for this block. They must be notified once
+	// the block is returned.
+	requests []*blockRetrievalRequest
+}
+
+// blockRetrievalQueue manages block retrieval requests. Higher priority
+// requests are executed first. Requests are executed in FIFO order within a
+// given priority level.
+type blockRetrievalQueue struct {
+	mtx sync.RWMutex
+	// queued or in progress retrievals
+	ptrs map[BlockPointer]*blockRetrieval
+	// global counter of insertions to queue
+	// capacity: ~584 years at 1 billion requests/sec
+	insertionCount uint64
+
+	heap *blockRetrievalHeap
+	// This is a channel of channels to maximize the time that each request is
+	// in the heap, allowing preemption as long as possible. This way, a
+	// request only exits the heap once a worker is ready.
+	workerQueue chan chan *blockRetrieval
+}
+
+// newBlockRetrievalQueue creates a new block retrieval queue. The numWorkers
+// parameter determines how many workers can concurrently call WorkOnRequest
+// (more than numWorkers will block).
+func newBlockRetrievalQueue(numWorkers int) *blockRetrievalQueue {
+	return &blockRetrievalQueue{
+		ptrs:        make(map[BlockPointer]*blockRetrieval),
+		heap:        &blockRetrievalHeap{},
+		workerQueue: make(chan chan *blockRetrieval, numWorkers),
+	}
+}
+
+// notifyWorker notifies workers that there is a new request for processing.
+func (brq *blockRetrievalQueue) notifyWorker() {
+	go func() {
+		// Get the next queued worker
+		ch := <-brq.workerQueue
```

Comment:
If there is no worker ready, this blocks -- do you think we'll ever need to clean up these blocked goroutines, say at the end of a test?  I can't think of a specific way this would cause problems, since it seems like every test _should_ wait for every block that was requested to finish.  But if you think there's a chance, we might want a `Shutdown` method to cancel any blocked goroutines.